src/Aspire.Hosting/Extensions/ResourceBuilderExtensions.cs

```diff
@@ -332,6 +332,30 @@ private static void ApplyEndpoints<T>(this IResourceBuilder<T> builder, IResourc
         return builder.WithAnnotation(annotation);
     }
 
+    /// <summary>
+    /// Changes an existing creates a new endpoint if it doesn't exist and invokes callback to modify the defaults.
+    /// </summary>
+    /// <param name="builder">Resource builder for resource with endpoints.</param>
+    /// <param name="endpointName">Name of endpoint to change.</param>
+    /// <param name="callback">Callback that modifies the endpoint.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<T> WithEndpoint<T>(this IResourceBuilder<T> builder, string endpointName, Action<EndpointAnnotation> callback) where T: IResourceWithEndpoints
+    {
+        var endpoint = builder.Resource.Annotations
+            .OfType<EndpointAnnotation>()
+            .Where(ea => StringComparers.EndpointAnnotationName.Equals(ea.Name, endpointName))
+            .SingleOrDefault();
+
+        if (endpoint == null)
+        {
+            endpoint = new EndpointAnnotation(ProtocolType.Tcp, name: endpointName);
+        }
```

Comment:
Feels like this should be optional.

Reply:
The protocol?

Reply:
The upsert behavior. What if I want to edit if it exists and noop if it doesn't.